.doom.d/config.org
* General

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Ben Whitley"
      user-mail-address "dev@purplg.com")
#+END_SRC

Disables the extra line at the end of file
#+BEGIN_SRC emacs-lisp
(setq require-final-newline nil)
#+END_SRC

Show the class/function location of the point in the file
#+BEGIN_SRC emacs-lisp
(which-function-mode)
#+END_SRC

This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type 'relative)
#+END_SRC


Save recent files every time a new file is 'found'
#+BEGIN_SRC emacs-lisp :tangle no
(add-hook! 'find-file-hook #'recentf-save-list)
#+END_SRC

Save recent files every 5 minutes
#+BEGIN_SRC emacs-lisp :tangle no
(run-at-time (current-time) (* 5 60) 'recentf-save-list)
#+END_SRC

Force minibuffers to follow the bottom of the output as the compilation (or test) runs. I wanted this functionality for running cargo tests so I could quickly see if a test passed or failed.
#+BEGIN_SRC emacs-lisp
(setq compilation-scroll-output t)
#+END_SRC

Garbage collection
#+BEGIN_SRC emacs-lisp
(setq gc-cons-threshold 402653184)
(setq gc-cons-percentage 0.6)
#+END_SRC

* Font
Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:

+ `doom-font'
+ `doom-variable-pitch-font'
+ `doom-big-font' -- used for `doom-big-font-mode'; use this for
  presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string. You generally only need these two:
setq doom-font (font-spec :family "monospace" :size 14))

#+BEGIN_SRC emacs-lisp :tangle no
(setq doom-font (font-spec :family "Source Code Pro" :size 14))
#+END_SRC

#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "Fira Code Retina" :size 12))
#+END_SRC

#+BEGIN_SRC emacs-lisp :tangle no
(setq doom-font (font-spec :family "Iosevka Term SS11" :size 14))
#+END_SRC
* Theme
There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:

#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-horizon)
#+END_SRC

** Active window highlight
change the background color of the [focused/active] window

There is a problem with this where if you have the same buffer open in multiple windows, both windows will be the inactive color despite one of them being active
#+BEGIN_SRC emacs-lisp :tangle no
(setq highlighted-window-face '(:background "#1a1c23"))
(defun highlight-selected-window ()
  "Highlight selected window with a different background color."
  (interactive)
  (walk-windows (lambda (w)
                  (unless (eq w (selected-window))
                    (with-current-buffer (window-buffer w)
                      (buffer-face-set highlighted-window-face)))))
  (buffer-face-set 'default))
(add-hook! 'buffer-list-update-hook 'highlight-selected-window)
#+END_SRC

* Key bindings
** Find file

#+BEGIN_SRC emacs-lisp
(map! :desc "File at point"
 :leader
 :n "f h"
 'find-file-at-point )
#+END_SRC

** LSP
#+BEGIN_SRC emacs-lisp
(defun lsp-import-and-format ()
  (interactive)
  (lsp-organize-imports)
  (lsp-format-buffer))

(map! :desc "LSP Format"
 :leader
 :n "c f"
 'lsp-import-and-format)
#+END_SRC

** Evil tweaks

Scroll the entire buffer up or down when ^ is held
#+BEGIN_SRC emacs-lisp
(defun scroll-page-line-down ()
  (interactive)
  (evil-scroll-line-down 1)
  (evil-next-line 1))

(defun scroll-page-line-up ()
  (interactive)
  (evil-scroll-line-up 1)
  (evil-next-line -1))

(map! :ni "^j" 'scroll-page-line-down)
(map! :ni "^k" 'scroll-page-line-up)
#+END_SRC

Clear all find matches
#+BEGIN_SRC emacs-lisp
(map! :desc "Clear search highlight"
 :leader
 :n "s c"
 'evil-ex-nohighlight )
#+END_SRC

Fold all folds at level 2... weird hack for some project I worked on a long time ago. Maybe try a different keybind or allow it to get an integer argument?
#+BEGIN_SRC emacs-lisp
(defun +fold/close-all-level-2 ()
  (interactive)
  (+fold/close-all 2))

(map!
 :leader
 :n "z m"
 '+fold/close-all-level-2)
#+END_SRC

#+BEGIN_SRC emacs-lisp
(map!
 :n "C-M-D"
 'evil-multiedit-match-all)
#+END_SRC

** Journal
#+BEGIN_SRC emacs-lisp
(map! :desc "Journal"
  :leader
  :n "j")

(map! :desc "New entry"
  :leader
  :n "j n"
  'org-journal-new-entry)

(map! :desc "Next journal"
  :leader
  :n "j l"
  'org-journal-next-entry)

(map! :desc "Previous journal"
  :leader
  :n "j h"
  'org-journal-previous-entry)
#+END_SRC

** Applications

#+BEGIN_SRC emacs-lisp
(map! :desc "Open terminal"
 :leader
 :n "o t"
 'eshell )

(map! :desc "Applications"
 :leader
 :n "a" )
#+END_SRC
** Point registers
#+BEGIN_SRC emacs-lisp
(map! :desc "Point to register"
      :leader
      :n "\""
      'point-to-register)

(map! :desc "Jump to point in register"
      :leader
      :n "'"
      'register-to-point)
#+END_SRC
** git-gutter

#+BEGIN_SRC emacs-lisp
(map! :desc "git hunk diff"
      :leader
      :n "g d"
      'git-gutter:popup-hunk)
#+END_SRC

* Packages
** persp-mode
Prevent creating new workspaces when opening new emacs-client
Found here: [[https://github.com/hlissner/doom-emacs/issues/1949][hlissner/doom-emacs#1949 {REQUEST} No new workspace created on incoming emacs...]]
#+BEGIN_SRC emacs-lisp
(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))
#+END_SRC
** which-key
#+BEGIN_SRC emacs-lisp
(setq which-key-idle-delay 0.3)
#+END_SRC
** org-mode

Set the default location of org files
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/.org/")
#+END_SRC

*** Roam
#+BEGIN_SRC emacs-lisp
(use-package! org-journal
  :custom
  (org-journal-date-prefix "#+TITLE: ")
  (org-journal-file-format "%Y-%m-%d.org")
  (org-journal-dir "~/.org-roam")
  (org-journal-date-format "%A, %d %B %Y"))
#+END_SRC

#+BEGIN_SRC emacs-lisp
(use-package! org-roam
  :commands (org-roam-insert org-roam-find-file org-roam)
  :init
  (setq org-roam-directory "~/.org")
  (setq org-roam-graph-viewer "/usr/bin/open")
;;  (map! :leader
;;        :prefix "n"
;;        :desc "Org-Roam-Insert" "i" #'org-roam-insert
;;        :desc "Org-Roam-Find"   "/" #'org-roam-find-file
;;        :desc "Org-Roam-Buffer" "r" #'org-roam)
  :config
  (org-roam-mode +1))
#+END_SRC
*** Capture Templates

#+BEGIN_SRC emacs-lisp :results silent
(setq +org-capture-todo-file "project.org")
(setq org-capture-templates
      '(("w" "Work"
         entry (file+headline "~/.org/Work.org" "Tasks")
         "* TODO %?\n %i\n")
        ("p" "Current project"
         entry (file+headline (lambda () (expand-file-name +org-capture-todo-file (projectile-project-root))) "Tasks")
        "* TODO %?\n%i\n%a" :prepend t)
        ("s" "Session"
         entry (file+headline "~/.org/PC.org" "Session")
        "* TODO %?\n%i" :prepend t)
        ("c" "PC"
         entry (file "~/.org/PC.org")
        "* TODO %?\n%i" :prepend t)
        ("h" "Home"
         entry (file+headline "~/.org/Home.org" "Tasks")
        "* TODO %?\n%i" :prepend t)))
#+END_SRC

** lsp-mode
*** General
#+BEGIN_SRC emacs-lisp
(after! lsp-mode
    :config
    (setq! lsp-enable-file-watchers nil))
#+END_SRC
*** Rust

Keybind to show UI doc overlay
#+BEGIN_SRC emacs-lisp
(after! lsp-mode
    :config
    (map! :desc "LSP show UI doc"
    :leader
    :n "c z"
    'lsp-ui-doc-show))
#+END_SRC

Enable by default
Currently not tangled because I prefer manually triggering with above keybind (SPC c z)
#+BEGIN_SRC emacs-lisp :tangle no
(add-hook! 'lsp-configure-hook 'lsp-ui-doc-mode)
#+END_SRC

** company-mode

#+BEGIN_SRC emacs-lisp
(after! company
  :config
  (setq! company-idle-delay 0
         company-minimum-prefix-length 3))
#+END_SRC

** projectile
https://github.com/ericdanan/counsel-projectile

This excludes a lot of files I never want to edit in emacs from the file finder.
Mostly just Unity stuff

#+BEGIN_SRC emacs-lisp :tangle no
(after! counsel
 (setq counsel-find-file-ignore-regexp "\\(?:^[#.]\\)\\|\\(?:[#~]$\\)\\|\\(?:^Icon?\\)\\|\\(?:.meta$\\)\\|\\(?:.asset$\\)\\|\\(?:.prefab$\\)"))
(after! counsel
  (append projectile-globally-ignored-directories '("~/.rustup/" "~/.cargo")))
#+END_SRC
** treemacs
https://github.com/Alexander-Miller/treemacs#projects-and-workspaces

Also excludes some Unity files from treemacs

#+BEGIN_SRC emacs-lisp
(after! treemacs
  :config
  (defun treemacs-ignore-unity (filename absolute-path)
    (or (string-suffix-p ".meta" filename t)
        (string-suffix-p ".asset" filename t)))
  (add-to-list 'treemacs-ignored-file-predicates #'treemacs-ignore-unity)
  (treemacs-follow-mode))
#+END_SRC

** deft
#+BEGIN_SRC emacs-lisp
(after! deft
;;  :bind
;;  ("C-c n d" . deft)
  (setq deft-recursive nil)
  (setq deft-use-filter-string-for-filename t)
  (setq deft-default-extension "org")
  (setq deft-directory "~/.org"))
#+END_SRC
** undo-fu

#+BEGIN_SRC emacs-lisp :tangle no
(add-hook 'evil-local-mode-hook 'turn-on-undo-tree-mode)
#+END_SRC
** hl-todo
Existing =hl-todo-keyword-faces=
#+BEGIN_SRC emacs-lisp :tangle no
(("TODO" warning bold)
 ("FIXME" error bold)
 ("HACK" font-lock-constant-face bold)
 ("REVIEW" font-lock-keyword-face bold)
 ("NOTE" success bold)
 ("DEPRECATED" font-lock-doc-face bold)
 ("BUG" error bold)
 ("XXX" font-lock-constant-face bold))
#+END_SRC

Append "LEARN" to above
#+BEGIN_SRC emacs-lisp
(after! hl-todo
    :config
    (add-to-list 'hl-todo-keyword-faces '("LEARN" font-lock-keyword-face bold)))
#+END_SRC
** telega

Keybind to open Telega, a Telegram client
#+BEGIN_SRC emacs-lisp
(map! :desc "Telega"
 :leader
 :n "o c"
 'telega )
#+END_SRC

These settings are required to be set manually when running emacs in daemon mode
#+BEGIN_SRC emacs-lisp
(setq telega-use-images t)
(setq telega-emoji-font-family "Noto Color Emoji")
(setq telega-emoji-use-images t)
(setq telega-online-status-function t)
#+END_SRC
** circe (irc)

Dunno why authentication fails when =:sasl_username= uses a similar closure as the =:sasl_secret= closure but for ~user~ instead. Whatever, the important part is the password
#+BEGIN_SRC emacs-lisp
(after! circe
  (set-irc-server! "irc.libera.chat"
    `(:tls t
      :port 6697
      :nick "purplg"
      :sasl-username "purplg"
      :sasl-password (lambda (&rest _) (+pass-get-secret "irc.libera.chat")))))
#+END_SRC
** rustic-mode

#+BEGIN_SRC emacs-lisp
(setq-hook! 'rustic-mode-hook fill-column 100)
#+END_SRC

*** ron-mode
#+BEGIN_SRC emacs-lisp
(use-package! ron-mode)
#+END_SRC
** code folding

#+BEGIN_SRC emacs-lisp :tangle no
(use-package! yafolding)
#+END_SRC


.doom.d/config.el
(setq user-full-name "Ben Whitley"
      user-mail-address "dev@purplg.com")
(setq require-final-newline nil)

(setq doom-font (font-spec :family "Source Code Pro" :size 14))

(setq doom-theme 'doom-horizon)

(setq display-line-numbers-type 'relative)

(map! :desc "File at point"
 :leader
 :n "f h"
 'find-file-at-point )

(defun lsp-import-and-format ()
  (interactive)
  (lsp-organize-imports)
  (lsp-format-buffer))

(map! :desc "LSP Format"
 :leader
 :n "c f"
 'lsp-import-and-format)

(defun scroll-page-line-down ()
  (interactive)
  (evil-scroll-line-down 1)
  (evil-next-line 1))

(defun scroll-page-line-up ()
  (interactive)
  (evil-scroll-line-up 1)
  (evil-next-line -1))

(map! :ni "^j" 'scroll-page-line-down)
(map! :ni "^k" 'scroll-page-line-up)

(map! :desc "Clear search highlight"
 :leader
 :n "s c"
 'evil-ex-nohighlight )

(defun +fold/close-all-level-2 ()
  (interactive)
  (+fold/close-all 2))

(map!
 :n "z m"
 '+fold/close-all-level-2)

(map! :desc "Journal"
  :leader
  :n "j")

(map! :desc "New entry"
  :leader
  :n "j n"
  'org-journal-new-entry)

(map! :desc "Next journal"
  :leader
  :n "j l"
  'org-journal-next-entry)

(map! :desc "Previous journal"
  :leader
  :n "j h"
  'org-journal-previous-entry)

(map! :desc "Open terminal"
 :leader
 :n "o t"
 'eshell )

(map! :desc "Applications"
 :leader
 :n "a" )

(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))

(setq which-key-idle-delay 0.3)

(setq org-directory "~/.org/")

(use-package! org-journal
  :custom
  (org-journal-date-prefix "#+TITLE: ")
  (org-journal-file-format "%Y-%m-%d.org")
  (org-journal-dir "~/.org-roam")
  (org-journal-date-format "%A, %d %B %Y"))

(use-package! org-roam
  :commands (org-roam-insert org-roam-find-file org-roam)
  :init
  (setq org-roam-directory "~/.org")
  (setq org-roam-graph-viewer "/usr/bin/open")
;;  (map! :leader
;;        :prefix "n"
;;        :desc "Org-Roam-Insert" "i" #'org-roam-insert
;;        :desc "Org-Roam-Find"   "/" #'org-roam-find-file
;;        :desc "Org-Roam-Buffer" "r" #'org-roam)
  :config
  (org-roam-mode +1))

(after! lsp-mode
    :config
    (setq! lsp-enable-file-watchers nil))

(after! company
  :config
  (setq! company-idle-delay 0
         company-minimum-prefix-length 3))

;(after! counsel
;  (setq counsel-find-file-ignore-regexp "\\(?:^[#.]\\)\\|\\(?:[#~]$\\)\\|\\(?:^Icon?\\)\\|\\(?:.meta$\\)\\|\\(?:.asset$\\)\\|\\(?:.prefab$\\)"))

(after! treemacs
  :config
  (defun treemacs-ignore-unity (filename absolute-path)
    (or (string-suffix-p ".meta" filename t)
        (string-suffix-p ".asset" filename t)))
  (add-to-list 'treemacs-ignored-file-predicates #'treemacs-ignore-unity)
  (treemacs-follow-mode))

(use-package! deft
  :after org
;;  :bind
;;  ("C-c n d" . deft)
  :custom
  (deft-recursive t)
  (deft-use-filter-string-for-filename t)
  (deft-default-extension "org")
  (deft-directory "~/.org"))
